KAFKA-4033: KIP-70: Revise Partition Assignment Semantics on New Consumer's Subscription Change

vahidhashemian opened the pull request:
This PR changes topic subscription semantics so a change in subscription does not immediately cause a rebalance.
Instead, the next poll or the next scheduled metadata refresh will update the assigned partitions.

vahidhashemian commented:
@hachikuji @ewencp I'd appreciate your initial feedback on this PR. The approach I took in handling topic subscriptions is simply mimicking how regex subscriptions are handled. Of course, this breaks a number of existing unit tests, but I'd like to have some feedback early on before diving into fixing those unit tests. Thanks a lot.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -808,7 +808,8 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
                 log.debug("Subscribed to topic(s): {}", Utils.join(topics, ", "));
                 this.subscriptions.subscribe(topics, listener);
-                metadata.setTopics(subscriptions.groupSubscription());
+                this.metadata.needMetadataForAllTopics(true);
```
Not sure I understand why this needs to be changed.

vahidhashemian replied:
@hachikuji Well the reason I removed the line `metadata.setTopics(subscriptions.groupSubscription());` is we make the same call in `addMetadataListener()` as part of the changes made in this PR ([here](https://github.com/vahidhashemian/kafka/blob/2a77e8c11bcc372f67a3dee95aa7f8eafbf1f3ec/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L166)). And I now realize we may not need the new line 811 above because, as I understand, we require fresh metadata for all topics when we are dealing with regex subscription. So, perhaps we just remove the line `this.metadata.needMetadataForAllTopics(true);` from this PR?

hachikuji replied:
Hmm... Yeah, maybe the call in the metadata listener is sufficient. But we definitely don't want to fetch metadata for all topics. Also, it seems unnecessary to request a metadata update blindly. I think `metadata.setTopics` was previously requesting the update only if the topics are not already contained in the Metadata.

vahidhashemian replied:
@hachikuji If I'm not mistaken you are suggesting to use this for metadata update:

```
                if (!this.metadata.topics().containsAll(topics))
                    this.metadata.requestUpdate();
```

The issue is if we just reduce the subscribed topics (e.g. from `[foo, bar, quux]` to `[foo, quux]`) with this code the metadata won't be updated until the next metadata refresh timeout (which is 5 minutes by default). Unless we use a check that verifies the two sets (metadata topics, and newly subscribed topics) are not the same?

hachikuji replied:
@vahidhashemian Are you concerned about the fact that we will have some unneeded metadata cached? That doesn't seem like a problem as long as we clean it up eventually (i.e. on the next scheduled refresh).

vahidhashemian replied:
@hachikuji The main concern based on some tests I ran is that in the above example the consumer is still subscribed to topic `bar` and receives its records until the next scheduled refresh (`bar` partitions will not revoke until the next scheduled refresh).

hachikuji replied:
Hmm.. But when you call `SubscriptionState.subscribe()`, we should set the flag saying that we need a rebalance. So the next call to poll() will force us to rebalance and we won't be able to return any data for an unsubscribed partition. Is that not right?

vahidhashemian replied:
@hachikuji I don't think that actually would work now, because we are not immediately updating subscribed topics anymore (the reason the KIP started). So even if we trigger a rebalance assigned partitions won't change until metadata gets updated. That tells me we need to update metadata in any case (perhaps with the exception of when subscribing to the exact same topics). Do you agree?

hachikuji replied:
When the user calls subscribe(), we should set a flag in SubscriptionState saying that a rebalance is needed. As of yet, the user's assignment hasn't changed. When the user calls poll(), we check the flag and begin the rebalance. Once the rebalance completes, we set the new assignment and begin fetching. I'm not sure I see a way for any fetches from the old assignment to reach the user.

hachikuji replied:
@vahidhashemian Feel free to ping me on Google chat if you want to talk about this. It would probably go a little faster that way.

vahidhashemian replied:
@hachikuji Thanks. I'm trying to consume, and update the code as per, your suggestion above. I'll reach out to you on Google if I need help.

vahidhashemian commented:
@hachikuji I made some changes according to your suggestion and it seems to work as expected now. I'll spend more time testing later on (I think I noticed more rebalances than actually required with this code), but I just updated the WIP PR so you can take a look too. Thanks.

vahidhashemian commented:
@hachikuji Correction: there is no unnecessary rebalances, just a redundant metadata update requests (with the test case I'm running) which I think is okay. And unit tests seem to be unaffected by the current PR (current build errors seem to be unrelated). Please let me know if you notice anything that I've missed.

vahidhashemian committed:
KAFKA-4033: KIP-70: Revise Partition Assignment Semantics on New Consumer's Subscription Change

This PR changes topic subscription semantics so a change in subscription does not immediately cause a rebalance.
Instead, the next poll or the next scheduled metadata refresh will update the assigned partitions.

vahidhashemian committed:
Unit test for subscription change

vahidhashemian commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -690,6 +691,112 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+
+        client.prepareResponseFrom(syncGroupResponse(Arrays.asList(partition2), Errors.NONE.code()), coordinator);
+
+        consumer.poll(0);
```
@hachikuji I am trying to write a few unit tests for KIP-70, but am running into some issues here. In this one, the subscription is set to `topic1` and then to `topic2`; but it seems that at this point I am not mocking some responses. Any idea? Thanks.

vahidhashemian committed:
Clean up KafkaConsumerTest.java and add reusable methods

vahidhashemian committed:
Fix unsubscribe semantics and add necessary unit tests

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -157,15 +157,23 @@ public void onMetadataUpdate(Cluster cluster) {
-                    for (String topic : cluster.topics())
-                        if (subscriptions.getSubscribedPattern().matcher(topic).matches() &&
-                                !(excludeInternalTopics && cluster.internalTopics().contains(topic)))
+                    for (String topic: topics)
```
nitpick: we usually include a space before and after ":"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -157,15 +157,23 @@ public void onMetadataUpdate(Cluster cluster) {
                             topicsToSubscribe.add(topic);
 
-                    subscriptions.subscribeFromPattern(topicsToSubscribe);
+                    if (hasAutoTopicSubscription)
```
I think I'm still not getting why we're trying to hook into the metadata updates in this patch. As I understand, the main point is to hold onto the current assignment metadata until the onPartitionsRevoked() hook. Unless I'm missing something, it shouldn't have anything to do with when we receive topic metadata (which can happen basically at any time).

vahidhashemian replied:
@hachikuji Yes, you are correct. I'll submit an update soon.

vahidhashemian replied:
@hachikuji Quick question. For `unsubscribe()` and manual `assign(...)` there is no following rebalance. Do you think a manual `commitAsync()` is required in these two cases to make sure offsets of to-be-revoked partitions are properly committed?

hachikuji replied:
@vahidhashemian Good question. For assign(), it makes sense to call commitAsync() prior to changing the assignment (arguably, it should be commitSync() but that would make assign() a blocking call). I think it also makes sense for unsubscribe since the current behavior is to remove the assignment immediately when it is called. Actually that seems reasonable since we also immediately leave the group (when using dynamic assignment).

vahidhashemian replied:
@hachikuji Thanks for confirming. I meant to mention it in my last message that I believe `commitAsync()` should be called in these two cases only when offsets are set to automatically be committed.

vahidhashemian commented:
@hachikuji I removed the WIP label from the PR. It would be great if you could take a look and share your feedback. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -872,6 +875,11 @@ public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
     public void unsubscribe() {
         acquire();
         try {
+            if (!this.subscriptions.allConsumed().isEmpty() && this.autoCommitEnabled)
```
Why is this check different from the one in `assign` below?

vahidhashemian replied:
I was thinking if there is no records consumed by the consumer (no `poll`) before calling `unsubscribe()` there is no need for a commit. I think we could change the check in `assign` method to this too. What do you think?

hachikuji replied:
I think we would end up skipping the offset commit anyway since we check for empty prior to sending it. In fact, we could probably simplify this to just:

``` java
if (this.autoCommitEnabled)
  this.commitAsync();
```

vahidhashemian replied:
Without the additional check one of the existing unit tests would fail ([this line](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java#L147)) I believe because there is no offset commit callback specified (which should be OK because there is no consumed offsets). What do you think?

hachikuji replied:
Interesting. So this seems to expose a couple problems. 
1. It looks like we allow the `OffsetCommitCallback` provided by the user to be null, so we probably need a null check in `ConsumerCoordinator.OffsetCommitCompletion.invoke()`.
2. For automatic commits, we use a logging `OffsetCommitCallback`. See `ConsumerCoordinator.maybeAutoCommitOffsetsAsync()`. We should probably use the same callback here, so maybe we need to make that method public  and invoke it instead of calling `commitAsync`.

vahidhashemian replied:
I am running into an issue with your suggestion in the second bullet. The method `maybeAutoCommitOffsetsAsync(...)` takes a `long`, and when I pass in `time.milliseconds()` (i.e. calling `this.coordinator.maybeAutoCommitOffsetsAsync(time.milliseconds());` here) it doesn't execute the main block of code because the condition `now >= nextAutoCommitDeadline` does not hold. It seems we need to pass in a `long` value that's greater than the next auto-commit time (in milliseconds). Since we just want to make sure the main block of code in that method (which involves committing the offsets) is executed, we could just pass `Long.MAX_VALUE`; though I'm not sure if that's the cleanest approach. What's your opinion?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -258,7 +258,10 @@ public void poll(long now) {
         pollHeartbeat(now);
-        maybeAutoCommitOffsetsAsync(now);
+        if (subscriptions.hasNoSubscription())
+            subscriptions.clearAssignment();
```
I'm wondering if it makes sense to move this to `onJoinPrepare` (after invocation of `onPartitionsRevoked`) since that is the last step before rejoining the group?

vahidhashemian replied:
I think I'm missing something. When calling `unsubscribe()` the method `onJoinPrepare` is not called (as it doesn't trigger a rebalance), so the assignment wouldn't get cleared. This would fail the added unit test `testSubscriptionChangesWithAutoCommitEnabled`. Could you please clarify?

hachikuji replied:
That's true, but perhaps that could be a call to `clearAssignment` directly from `unsubscribe`? If we only clear the assignment here, then it wouldn't be cleared until the user calls `poll` after calling `unsubscribe` which doesn't seem quite right.

vahidhashemian replied:
Yes, I was debating that myself. Then I thought if a consumer is subscribed to topic `foo`, the behavior should be the same (speaking of assigned partitions) in the following scenarios that would happen next:
- calling `subscribe(Collections.singleton("bar"))`
- calling `unsubscribe()` first, and then `subscribe(Collections.singleton("bar"))`

If we clear the assignment right away the above use cases would lead to slightly different behavior. Unless that's Okay.

hachikuji replied:
I think that's fine. As I see it, we should always reflect the active assignment. When we call `unsubscribe`, we leave the group, so the current assignment is no longer active. On the other hand, if we don't call `unsubscribe`, then the assignment will be active until the `onPartitionsRevoked` callback is triggered. Does that make sense or not?

vahidhashemian replied:
Thanks. Yes it makes sense, I just wanted to be clear about the `unsubscribe()` partition assignment semantics. I'll update the patch so that assigned partitions are also cleared immediately when calling `unsubscribe()`. The only update to `unsubscribe()` would be the call to `commitAsync()` if auto-commit enabled and there are consumed records.

ijuma commented:
@hachikuji, this looks like a good candidate for your first KIP merge :)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -128,13 +127,6 @@ private void changeSubscription(Set<String> topicsToSubscribe) {
             this.groupSubscription.addAll(topicsToSubscribe);
-
-            // Remove any assigned partitions which are no longer subscribed to
-            for (Iterator<TopicPartition> it = assignment.keySet().iterator(); it.hasNext(); ) {
```
I'm surprised that this didn't cause any unit test failures in `SubscriptionStateTest`. Haha, that probably shows that we have some gaps in our coverage. Maybe we could have test cases which validate that the assignment does not change with a call to `subscribe`, but does when we call `assignFromSubscribed`? Another one for regex would be nice as well. These cases are probably covered in `KafkaConsumerTest` already, but it's helpful to have unit testing as close as possible to the implementation.

vahidhashemian replied:
I believe this method is not subsequently called in `assignFromSubscribed` or `assignFromUser`. Plus, it seems there are already existing unit tests in `SubscriptionStateTest` that verify that the assignment is updated only when necessary. Here are examples for [`assignFromUser`](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L49), [`assignFromSubscribed`](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L92), [`subscribe`](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L82), and [`unsubscribe`](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L190). There are a couple missing for pattern `subscribe` and `subscribeFromPattern`, that I could add after [this line](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L158) and [this line](https://github.com/vahidhashemian/kafka/blob/f60fb52a702a6bcb854e9560548174fc5ad86730/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L159). Let me know if I'm missing further unit tests that are required. Thanks.

hachikuji replied:
What I'm looking for is test cases which would have caused this code deletion to fail. Perhaps the main case we're missing is one which asserts that the assignment remains the same after a subsequent call to `subscribe()`, and only changes when `assignFromSubscribed` is called. Glancing at the code, I'm also not sure we have a case which covers a subsequent call to `subscribe(Pattern)` as well.

vahidhashemian replied:
Sounds good. I added a unit test in `SubscriptionStateTest`. I hope that covers what you described above.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -60,6 +61,49 @@ public void partitionAssignment() {
+        assertTrue(state.assignedPartitions().isEmpty());
+
+
+        state.subscribe(Pattern.compile(".*"), rebalanceListener);
```
nit: would you mind splitting from here on into a separate test case specific to patterns? The method name serves as documentation and also isolates the functionality under test a little better which makes debugging easier.

vahidhashemian replied:
Sure, I'll do that in the next round.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -872,6 +875,11 @@ public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
     public void unsubscribe() {
         acquire();
         try {
+            if (this.autoCommitEnabled)
```
Looks like `maybeAutoCommitOffsetsAsync` already checks this flag.

vahidhashemian replied:
That's right. I'll remove this check. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -899,6 +907,12 @@ public void unsubscribe() {
     public void assign(Collection<TopicPartition> partitions) {
         acquire();
+
+        if (!this.assignment().isEmpty() && this.autoCommitEnabled)
```
This should probably be replaced with a call to `maybeAutoCommitOffsetsAsync` as well, right?

vahidhashemian replied:
Both suggestions make sense. I'll update the code to reflect these. `maybeAutoCommitOffsetAsync` commits offsets for `subscriptions.allConsumed()` which should cover the above check `!this.assignment().isEmpty()`. If nothing has been consumed no commit is made.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -899,6 +907,12 @@ public void unsubscribe() {
     public void assign(Collection<TopicPartition> partitions) {
         acquire();
+
+        if (!this.assignment().isEmpty() && this.autoCommitEnabled)
```
One other thing. Would it make sense to move this call to the `else` branch below (just before the call to `assignFromUser`)? Otherwise, we'd double-commit down the `unsubscribe` path.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -507,6 +507,7 @@
     private final Metadata metadata;
     private final long retryBackoffMs;
     private final long requestTimeoutMs;
+    private final boolean autoCommitEnabled;
```
Can we get rid of this field now?

vahidhashemian replied:
This field is still useful for unit tests (in `KafkaConsumerTest`) that need consumers with auto-commit enabled or disabled. Currently, we seem to be passing the flag to the `KafkaConsumer` constructor but not using it. Unless you think we should handle those scenarios differently (e.g. using `ConsumerConfig`)?

hachikuji replied:
Hmm... But the value is not actually read anywhere, right? Wouldn't the tests just check that `maybeAutoCommitOffsetsAsync` gets invoked (regardless of this field)?

vahidhashemian replied:
Yes, that's right. The flag needs to be passed to the coordinator, which should work without the member variable. Updating the patch now ... Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -872,6 +875,10 @@ public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
         try {
+            // make sure the offsets of topic partitions the consumer is unsubscribing from
+            // are committed since there will be no following rebalance
+            this.coordinator.maybeAutoCommitOffsetsAsync(Long.MAX_VALUE);
```
Maybe I missed a comment, but why not use `time.milliseconds()`? It might even make sense to have the public `maybeAutoCommitOffsetsAsync` accept no arguments.

vahidhashemian replied:
Sorry, I made a comment on this yesterday but became hidden after I submitted an update. Here's the comment:

> I am running into an issue with your suggestion in the second bullet. The method maybeAutoCommitOffsetsAsync(...) takes a long, and when I pass in time.milliseconds() (i.e. calling this.coordinator.maybeAutoCommitOffsetsAsync(time.milliseconds()); here) it doesn't execute the main block of code because the condition now >= nextAutoCommitDeadline does not hold. It seems we need to pass in a long value that's greater than the next auto-commit time (in milliseconds). Since we just want to make sure the main block of code in that method (which involves committing the offsets) is executed, we could just pass Long.MAX_VALUE; though I'm not sure if that's the cleanest approach.

If we create a public `maybeAutoCommitOffsetAsync` that calls the old private method we would still need to pass in the `long` value. Unless the public one repeats the code inside the private method without the timestamp handling?

hachikuji replied:
Aha, I missed the significance of the parameter. We actually want the commit to happen regardless of the next auto commit deadline. So maybe we can refactor as you suggested and have the public version accept no arguments and simply send the offset commit. Maybe we should call it `maybeAutoCommitNow` or something like that? Then the existing method can delegate to this method when the commit deadline has been reached.

vahidhashemian replied:
Sounds good. I'll add it in the next patch.

vahidhashemian commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -514,22 +514,32 @@ private void maybeAutoCommitOffsetsAsync(long now) {
+     * @param bypassChecks: whether the auto-commit and coordinator status can be
+     * bypassed or have to be double-checked
+     */
+    public void maybeAutoCommitOffsetsNow(boolean bypassChecks) {
```
@hachikuji I added a boolean parameter so we don't re-evaluate the checks already performed in the existing private method. Please advise if you think we should go about this in a different way. Thanks.

hachikuji replied:
The extra parameter seems a little more annoying than the effort it saves (which is really just checking a couple flags). If we really wanted to avoid the redundant checks, maybe a better way is to add a private `doAutoCommitAsync` or something like that, which simply sends the request assuming the coordinator is known and autocommit is enabled. Then the two `maybeAutoCommit` calls could delegate to `doAutoCommitAsync` after doing the checks themselves.

vahidhashemian replied:
That's fair. I'll update it like that.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -60,6 +61,51 @@ public void partitionAssignment() {
+        // assigned partitions should immediately change
+        assertEquals(singleton(tp1), state.assignedPartitions());
+
+        state.unsubscribe();
```
Prior to unsubscribing, can we verify the assigned partitions remain the same with another call to `subscribeFromPattern`?

vahidhashemian replied:
Sure.

vahidhashemian commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -706,10 +706,10 @@ private KafkaConsumer(ConsumerConfig config,
                   int heartbeatIntervalMs,
                   long retryBackoffMs,
+                  boolean autoCommitEnabled,
                   long requestTimeoutMs) {
```
We could also remove the unused parameters from this constructor.

hachikuji replied:
True. I went ahead and did that in KAFKA-3807. I'm not sure which will land first, but feel free to do it here.

vahidhashemian replied:
Sure, one of us will have to rebase :)

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -60,6 +61,51 @@ public void partitionAssignment() {
+
+        state.subscribe(singleton(topic1), rebalanceListener);
+        // assigned partitions should remain unchanged
+        assertTrue(state.assignedPartitions().isEmpty());
```
This doesn't seem too interesting as a test case since we already unsubscribed. Maybe the second `subscribe` can go before the `unsubscribe`? Then we can verify that we still own tp0 and tp1.

vahidhashemian replied:
The problem is in that case we run into the exception: `java.lang.IllegalStateException: Subscription to topics, partitions and pattern are mutually exclusive`. Unless I misunderstood?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -60,6 +61,51 @@ public void partitionAssignment() {
+        state.unsubscribe();
+        // assigned partitions should immediately change
+        assertTrue(state.assignedPartitions().isEmpty());
+
```
nit: unneeded newline

vahidhashemian replied:
Got it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -706,10 +706,10 @@ private KafkaConsumer(ConsumerConfig config,
                   int autoCommitIntervalMs,
                   int heartbeatIntervalMs,
                   long retryBackoffMs,
+                  boolean autoCommitEnabled,
```
nit: the previous location seemed a little better since it was listed next to `autoCommitIntervalMs`.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -634,16 +783,130 @@ public void testWakeupWithFetchDataAvailable() {
+     * are not automatically committed (when auto-commit is disabled).
+     */
+    @Test
+    public void testManualSubscriptionWithAutoOffsetCommitDisabled() {
```
nit: usually we us "subscription" to refer to topics and "assignment" when referring to partitions. Maybe a better name is something like `testManualAssignmentChangeWithAutoCommitDisabled`. Then perhaps the comment above is unneeded?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -634,16 +783,130 @@ public void testWakeupWithFetchDataAvailable() {
+        int sessionTimeoutMs = 30000;
+        int heartbeatIntervalMs = 3000;
+
+        // adjust auto commit interval lower than heartbeat so we don't need to deal with
```
Heartbeats shouldn't come into play with manual assignment, so we can get rid of this comment.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -563,67 +499,280 @@ public void testAutoCommitSentBeforePositionUpdate() {
+     * partitions are committed (when auto-commit is enabled).
+     */
+    @Test
+    public void testManualSubscriptionWithAutoOffsetCommitEnabled() {
```
Similar to below. Maybe `testManualAssignmentChangeWithAutoOffsetCommitEnabled` is a more descriptive name.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -563,67 +499,280 @@ public void testAutoCommitSentBeforePositionUpdate() {
+        int sessionTimeoutMs = 30000;
+        int heartbeatIntervalMs = 3000;
+
+        // adjust auto commit interval lower than heartbeat so we don't need to deal with
```
We can remove this comment here also since auto-commit is disabled.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -563,67 +499,280 @@ public void testAutoCommitSentBeforePositionUpdate() {
+
+        // subscription change
+        consumer.subscribe(Arrays.asList(topic2), getConsumerRebalanceListener(consumer));
+
```
This is just indirect testing of `SubscriptionState`, but could you assert after the second `subscribe` that the assignment is still tp0?

vahidhashemian replied:
Sure, I'll add checks after this and the following `unsubscribe`.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -717,13 +1023,29 @@ private Struct listOffsetsResponse(Map<TopicPartition, Long> offsets, short erro
+        return response.toStruct();
+    }
+
+    private Struct fetchResponse(TopicPartition partition, long fetchOffset, int count) {
```
Can this be implemented just by delegating to the generic version with a singleton map?

vahidhashemian replied:
Yes, good suggestion.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -655,39 +907,82 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
 
-        // make sure the position hasn't been updated
-        assertEquals(0, consumer.position(partition));
+    private AtomicBoolean prepareOffsetCommitResponse(MockClient client, Node coordinator, final TopicPartition partition, final long offset) {
```
Here too it seems like we can use the generic version of `prepareOffsetCommitResponse`.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -60,6 +61,54 @@ public void partitionAssignment() {
+        // assigned partitions should remain unchanged
+        assertEquals(singleton(tp1), state.assignedPartitions());
+
+        state.unsubscribe();
```
Maybe just one more case missing. If we call subscribe with a new pattern, the assignment also shouldn't change.

vahidhashemian replied:
Sure, I'll add it.

vahidhashemian replied:
And will do the same for the topic subscription unit test.

hachikuji commented:
LGTM. Thanks for patch! Committed to trunk.
